packages/react-reconciler/src/__tests__/useEvent-test.internal.js

```diff
@@ -0,0 +1,595 @@
+  }
+
+  function Text(props) {
+    Scheduler.unstable_yieldValue(props.text);
+    return <span prop={props.text} />;
+  }
+
+  // @gate enableUseEventHook
+  it('memoizes basic case correctly', () => {
+    class IncrementButton extends React.PureComponent {
+      increment = () => {
+        this.props.onClick();
+      };
+      render() {
+        return <Text text="Increment" />;
+      }
+    }
+
+    function Counter({incrementBy}) {
+      const [count, updateCount] = useState(0);
+      const onClick = useEvent(() => updateCount(c => c + incrementBy));
+
+      return (
+        <>
+          <IncrementButton onClick={onClick} ref={button} />
+          <Text text={'Count: ' + count} />
+        </>
+      );
+    }
+
+    const button = React.createRef(null);
+    ReactNoop.render(<Counter incrementBy={1} />);
+    expect(Scheduler).toFlushAndYield(['Increment', 'Count: 0']);
+    expect(ReactNoop.getChildren()).toEqual([
+      span('Increment'),
+      span('Count: 0'),
+    ]);
+
+    act(button.current.increment);
+    expect(Scheduler).toHaveYielded([
+      // Button should not re-render, because its props haven't changed
+      'Count: 1',
+    ]);
+    expect(ReactNoop.getChildren()).toEqual([
+      span('Increment'),
+      span('Count: 1'),
+    ]);
+
+    act(button.current.increment);
+    expect(Scheduler).toHaveYielded([
+      // Event should use the updated callback function closed over the new value.
+      'Count: 2',
+    ]);
+    expect(ReactNoop.getChildren()).toEqual([
+      span('Increment'),
+      span('Count: 2'),
+    ]);
+
+    // Increase the increment prop amount
+    ReactNoop.render(<Counter incrementBy={10} />);
+    expect(Scheduler).toFlushAndYield(['Count: 2']);
+    expect(ReactNoop.getChildren()).toEqual([
+      span('Increment'),
+      span('Count: 2'),
+    ]);
+
+    // Event uses the new prop
+    act(button.current.increment);
+    expect(Scheduler).toHaveYielded(['Count: 12']);
+    expect(ReactNoop.getChildren()).toEqual([
+      span('Increment'),
+      span('Count: 12'),
+    ]);
+  });
+
+  // @gate enableUseEventHook
+  it('throws when called in render', () => {
+    class IncrementButton extends React.PureComponent {
+      increment = () => {
+        this.props.onClick();
+      };
+
+      render() {
+        // Will throw.
+        this.props.onClick();
+
+        return <Text text="Increment" />;
+      }
+    }
+
+    function Counter({incrementBy}) {
+      const [count, updateCount] = useState(0);
+      const onClick = useEvent(() => updateCount(c => c + incrementBy));
+
+      return (
+        <>
+          <IncrementButton onClick={onClick} />
+          <Text text={'Count: ' + count} />
+        </>
+      );
+    }
+
+    ReactNoop.render(<Counter incrementBy={1} />);
+    expect(Scheduler).toFlushAndThrow(
+      'An event from useEvent was called during render',
+    );
+
+    // TODO: Why?
+    expect(Scheduler).toHaveYielded(['Count: 0', 'Count: 0']);
```

Comment:
I'm not really sure what's going on here - cc @acdlite @rickhanlonii if you have any pointers on where to debug

Reply:
My question here was why/how we rendered twice but I think the answer is that we retry errors?

Also, I think we need to rewrite this test since it tests passing down the event function (what are we calling this?) which is not going to be supported in this version.

Reply:
@acdlite do you have a code pointer for where that works so I can step through how it happens? Just want to understand how it works better.

Reply:
recoverFromConcurrentError in the work loop